x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticator.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc.jwt;
+
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig;
+import org.elasticsearch.xpack.core.security.authc.jwt.JwtRealmSettings;
+import org.elasticsearch.xpack.core.ssl.SSLService;
+
+import java.text.ParseException;
+import java.time.Clock;
+import java.util.List;
+
+public class JwtAuthenticator implements Releasable {
+
+    private static final Logger logger = LogManager.getLogger(JwtAuthenticator.class);
+    private final RealmConfig realmConfig;
+    private final List<JwtClaimValidator> jwtClaimValidators;
+    private final JwtSignatureValidator jwtSignatureValidator;
+
+    public JwtAuthenticator(
+        final RealmConfig realmConfig,
+        final SSLService sslService,
+        final JwtSignatureValidator.PkcJwkSetReloadNotifier reloadNotifier
+    ) {
+        this.realmConfig = realmConfig;
+        final TimeValue allowedClockSkew = realmConfig.getSetting(JwtRealmSettings.ALLOWED_CLOCK_SKEW);
+        final Clock clock = Clock.systemUTC();
+        jwtClaimValidators = List.of(
+            new JwtHeaderValidator(realmConfig.getSetting(JwtRealmSettings.ALLOWED_SIGNATURE_ALGORITHMS)),
```

Comment:
I think ideally we could aggregate all validation errors (except signature) before returnning back the caller. This would help fixing any issues faster. But for now I don't change the logic here. But this small reordering is viable. So I moved it after validations of other claims.

Reply:
I noticed that existing code has type header check first, then `iss`, `aud`, then it is algorithm. If this order is still desirable, I can change the code to fit for it (basically split type and algorithm into 2 validators and place them in the right positions among the claim validators). Please let me know.

Reply:
Splitting the header validation seems like the right order to me.

Reply:
I split the header validation into two validators. Both header and cliam validators are now implemenations of `JwtFieldValidator` so that they can exist in a single list with the right order.

I also added a new test class ([JwtAuthenticatorTests](https://github.com/elastic/elasticsearch/pull/91001/files#diff-8da074ef8e1309103eec64df37abae77919b528cf7f8092e04e9652c8e5513e8)) for this particular order requirement. The class also shows how future tests can be mostly written in a more lightweight way without having to build the realm and create actual signed JWT tokens.